registry/remote/repository.go

```diff
@@ -660,6 +673,61 @@ func (s *blobStore) Fetch(ctx context.Context, target ocispec.Descriptor) (rc io
 	}
 }
 
+// Mount mounts the given descriptor from fromRepo into s.
+func (s *blobStore) Mount(ctx context.Context, desc ocispec.Descriptor, fromRepo string) error {
+	// pushing usually requires both pull and push actions.
+	// Reference: https://github.com/distribution/distribution/blob/v2.7.1/registry/handlers/app.go#L921-L930
+	ctx = registryutil.WithScopeHint(ctx, s.repo.Reference, auth.ActionPull, auth.ActionPush)
+
+	// We also need pull access to the source repo.
+	fromRef := s.repo.Reference
+	fromRef.Repository = fromRepo
+	ctx = registryutil.WithScopeHint(ctx, fromRef, auth.ActionPull)
+
+	url := buildRepositoryBlobMountURL(s.repo.PlainHTTP, s.repo.Reference, desc.Digest, fromRepo)
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, nil)
+	if err != nil {
+		return err
+	}
+	resp, err := s.repo.client().Do(req)
+	if err != nil {
+		return err
+	}
+	resp.Body.Close()
+	if resp.StatusCode == http.StatusCreated {
+		return nil
+	}
+	if resp.StatusCode != http.StatusAccepted {
+		return errutil.ParseErrorResponse(resp)
+	}
+	// From the spec:
+	//
+	// "If a registry does not support cross-repository mounting
+	// or is unable to mount the requested blob,
+	// it SHOULD return a 202.
+	// This indicates that the upload session has begun
+	// and that the client MAY proceed with the upload."
+	//
+	// So we need to pull the content in order to push it.
```

Comment:
It seems questionable, and now I fully understand you were referring in the [#oras slack channel thead](https://cloud-native.slack.com/archives/CJ1KHJM5Z/p1683220554237559).

In the most scenarios (like `docker push`), the blob source is locally available and should not be fetched again from the remote registry, which is inefficient.

By following your idea from the slack channel, it is more reasonable to add a parameter to `Mount()` similar to `GetBody func() (io.ReadCloser, error)` from [net/http.Request](https://pkg.go.dev/net/http#Request).

```go
func Mount(ctx context.Context, desc ocispec.Descriptor, fromRepo string, getContent func() (io.ReadCloser, error)) error
```
where `getContent` can be `nil`. In that case, an error will be returned on `202 Accepted`.

Reply:
The caller site might be
```go
err := repo.Mount(ctx, desc, "hello-world", func() (io.ReadCloser, error) {
    return os.Open("layer.tar.gz")
})
if err != nil {
    return err
}
```

Reply:
> In that case, an error will be returned on 202 Accepted.

I would actually suggest that when `getContent` is nil, we should do the pull/push thing because that's a useful default fallback behaviour that can work in all cases to accomplish what the caller is aiming to do.

If the user really doesn't want that to happen, they can still get the error-returning behaviour by passing in a function that always returns an error.

Reply:
I've gone with that approach, as I think it's clearly more useful behaviour. Happy to change it if you still think that's not the right way.

Reply:
I still think that should not pull / push things if `getContent` is `nil` because it eliminates an option that the users might just want to mount and fail directly if it is not mountable.